lib/grid_search.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python
+"""
+A module that provides functionalities for grid search
+will be used for hyperparameter optimization
+"""
+
+from lib import collaborative_filtering as cf
+from util import data_parser as dp
+from lib import evaluator as ev
+import numpy
+import itertools as it
+import sys
+import os
+sys.path.insert(1, os.path.join(sys.path[0], '..'))
+
+
+class GridSearch(object):
+
+    def __init__(self, recommender, hyperparameters):
+        """
+        Train number of recommenders using UV decomposition
+        using different parameters.
+        @param (object) recommender
+        @param (dict) hyperparameters, list of the hyperparameters
+        """
+        self.recommender = recommender
+        self.hyperparameters = hyperparameters
+        self.evaluator = ev.Evaluator(recommender.get_ratings())
+
+    def get_all_permutations(self):
+        """
+        the method retuns all possible permutations
+        of the hyperparameters
+        @returns (dict[]) array of dicts containing all
+        permutations
+        """
+        names = sorted(self.hyperparameters)
+        return [dict(zip(names, prod)) for prod in it.product(
+            *(self.hyperparameters[name] for name in names))]
+
+    def train(self):
+        """
+        The method loops on all hyperparameters and calls
+        the train and split method on the recommender then
+        calls the get_rmse from the evaluator. it returns
+        the best hyperparameters.
+        """
+        best_error = numpy.inf
+        keys = list(self.hyperparameters.keys())
+        best_params = dict()
+        train, test = self.recommender.split()
+        for config in self.get_all_permutations():
```

Comment:
This is the whole idea of the grid_search, some permutations report high testing error and low training one, we need to find the hyperparameters that minimizes the test error for the same splitted data.